# Citation

Nanopype integrates a growing number of nanopore sequencing related tools. If you use nanopype in your project please cite our pipeline. Furthermore please verify the set of tools your analysis made use of and also cite the according publications.


## Nanopype

>Pay Giesselmann, Sara Hetzel, Franz-Josef Müller, Alexander Meissner, Helene Kretzmer, Nanopype: a modular and scalable nanopore data processing pipeline. Bioinformatics, doi: 10.1093/bioinformatics/btz461 (2019).

## Snakemake

>J. Köster and S. Rahmann; Snakemake - A scalable bioinformatics workflow engine, Bioinformatics (2012).

## Processing functions

####**Minimap2**
>H. Li; Minimap2: pairwise alignment for nucleotide sequences. Bioinformatics. doi:10.1093/bioinformatics/bty191 (2018).

####**GraphMap**
>Sovic, I. et al. Fast and sensitive mapping of nanopore sequencing reads with GraphMap. Nat. Commun. 7:11307 doi: 10.1038/ncomms11307 (2016).

####**NGMLR**
>Sedlazeck, F. J. et al. Accurate detection of complex structural variations using single-molecule sequencing. Nature Methods 15, 461-468 (2018).

## Analysis functions

####**Nanopolish**
>Simpson, Jared T. et al. Detecting DNA cytosine methylation using nanopore sequencing. Nature Methods 14.4, 407-410 (2017).

####**Deepbinner**
>Wick, R. R., Judd, L. M. & Holt, K. E; Deepbinner: Demultiplexing barcoded Oxford Nanopore reads with deep convolutional neural networks, PLOS Computational Biology 14, e1006583 (2018).

####**Sniffles**
>Sedlazeck, F. J. et al. Accurate detection of complex structural variations using single-molecule sequencing, Nature Methods 15, 461-468 (2018).

## Miscellaneous
####**Bedtools**
>Aaron R. Quinlan, Ira M. Hall; BEDTools: a flexible suite of utilities for comparing genomic features, Bioinformatics, doi:10.1093/bioinformatics/btq033 (2010).

####**Samtools**
>Heng Li, Bob Handsaker, Alec Wysoker, Tim Fennell, Jue Ruan, Nils Homer, Gabor Marth, Goncalo Abecasis, Richard Durbin, 1000 Genome Project Data Processing Subgroup; The Sequence Alignment/Map format and SAMtools, Bioinformatics, doi:10.1093/bioinformatics/btp352 (2009).

####**UCSCTools**
>W. J. Kent, A. S. Zweig, G. Barber, A. S. Hinrichs, D. Karolchik; BigWig and BigBed: enabling browsing of large distributed datasets, Bioinformatics, doi:10.1093/bioinformatics/btq351 (2010).
